webhooks: Improve how we extract and validate HTTP event type headers.

eeshangarg opened the pull request:
@timabbott: FYI! Actually, one thing I am not clear on is what should we return when we have failed to find an event header, should we return `json_success` or let the webhook code raise whatever error will be caused by not having the relevant header in the first place? Thanks!

This PR relates to #6213! :) Thanks!

timabbott commented on zerver/webhooks/github/view.py:
```diff
@@ -371,15 +372,17 @@ def api_github_webhook(
+def get_event(request: HttpRequest, user_profile: UserProfile,
+              payload: Dict[str, Any], branches: Text) -> Optional[str]:
+    event = validate_extract_webhook_http_header(
+        'HTTP_X_GITHUB_EVENT', request, user_profile)
```
I think a function name and argument structure that's more like this:

`event = extract_header(request, HTTP_X_GITHUB_EVENT)` 

would be cleaner; you can get the user from `request.user`.

timabbott commented on zerver/lib/webhooks/common.py:
```diff
@@ -32,3 +39,18 @@ def check_send_webhook_message(
+            get_system_bot(settings.NOTIFICATION_BOT),
+            user_profile.bot_owner,
+            message_body
+        )
```
Do we want to raise a `JsonableError` here, so that callers don't need to handle the case where this returns `None`?

eeshangarg commented:
@timabbott: Updated! :) Thanks for the prompt review!

eeshangarg commented:
@timabbott: Also added a commit fixing a user-reported bug in our `bitbucket2` integration. :)

timabbott commented on zerver/webhooks/groove/tests.py:
```diff
@@ -108,12 +108,6 @@ def test_groove_ticket_state_changed(self) -> None:
-        self.subscribe(self.test_user, self.STREAM_NAME)
-        result = self.client_post(self.url, self.get_body('ticket_state_changed'),
-                                  content_type="application/x-www-form-urlencoded")
-        self.assert_json_error(result, 'Missing event header')
```
I think we actually do want a test of this form on at least one of the webhooks using this new header extraction feature; this sort of end-to-end test helps protect against situations where the unit tests work, but the function doesn't.

timabbott replied:
It might be reasonable to just do an end-to-end style test in the unit tests for the function (one that does an HTTP request into a webhook, not making a mock)

(Read about this high-level design concept in http://web.mit.edu/Saltzer/www/publications/endtoend/endtoend.pdf)

timabbott commented on zerver/webhooks/bitbucket2/view.py:
```diff
@@ -110,7 +113,6 @@ def get_subject_based_on_type(payload: Dict[str, Any], type: str) -> Text:
     return get_subject(payload)
 
 def get_type(request: HttpRequest, payload: Dict[str, Any]) -> str:
-    event_key = request.META.get("HTTP_X_EVENT_KEY")
```
Can you explain why moving this further down is the right thing to do?  

I feel like this is a better thing to validate first, both to avoid duplicated code, and also because this is sorta an initial sanity check that the event is intended for the bitbucket webhook...

eeshangarg replied:
@timabbott: Well, moving it further down makes sense here because in the bitbucket2 webhook, we don't always rely on the header to figure out the event type, so it makes sense to only get the header when needed. I made a comment about this further down! :) We might want to change that somewhere along the way, but that should be its own PR!

timabbott replied:
Are there payloads where they don't send the event type header at all?  Regardless of how we do our parsing, I think it makes sense to fail here with a clear exception about the header being missing, rather than failing further on because of one of the other checks to determine event type.

eeshangarg replied:
Okay, so there aren't any payloads that don't send the event type header. It's just that we choose not to rely on the header for most part. I have no idea why that is, the person who wrote this integration just chose to figure out the event type from the payload JSON itself and not from the header (which would have been easier)! The way I see it, we have two options:

* Keep the status quo and simply raise an exception if the header is missing, like you suggested. This would mean passing in the correct headers in the tests, which isn't the case right now, since the header isn't relied upon to determine the even type.
* The other option is to rewrite `get_type` from scratch and make sure that we use the event header to determine the event type. This is a larger project we should eventually get to, but isn't super high priority.

What do you think? :)

timabbott replied:
I think option 1 is fine for now; it's kinda the first step regardless.

timabbott commented on zerver/lib/webhooks/common.py:
```diff
@@ -1,13 +1,33 @@
+Hi there! We thought you'd like to know that the integration associated
+with your bot {bot_name} just received a payload that is missing the
+HTTP header that allows us to determine the event type.
+"""
```
This error message needs to be redone.  Think about the context of a user receiving this: You just setup your GitLab webhook and you get this error.  What are you supposed to do?

I think we want something like this:
```
Hi there!  Your bot {bot_name} just sent an HTTP request to {request.path} that is missing the 
HTTP {header_name} header.  Because this header is how {integration_name} indicates the event type,
this usually indicates a configuration issue, where you either entered the URL for a different integration.  
Contact {support_email} if you need help debugging!
```

Where `header_name` has the Django-specific `HTTP_` stripped from the front (possibly, rather than parsing, we should have the webhook code pass in `X_GITHUB_EVENT` and have the library code add the `HTTP_` on the end; I think that would simplify how this works for webhook authors).  

(Also, we should document this new function in the webhook walkthrough).

timabbott commented:
I posted a few code things.  Also, we should:
(1) Use the `last_reminder` logic for rate-limiting these, to avoid sending a user dozens of these in a short time.  I would probably do that by extracting a function in a commit at the start of the branch that we can call.
(2) Document this new function in the webhook walkthrough.

eeshangarg commented:
@timabbott: Updated this, however, this is not done yet. Here are a couple of things I plan on tackling tomorrow:
* Document the function in the webhook walkthrough.
* Use the `last_reminder` logic for rate-limiting these messages.
* Also, the `bitbucket2` stuff I commented about above.

Here are few things I did end up doing:
* Added a test that tested the header extraction function with an actual webhook, instead of making a mock request.
* Changed the notification message to be more informational, using the message template you came up with.

Of course, I do not expect this to be merged immediately, this turned into quite a large PR. But I would appreciate your feedback on the work I've pushed so far, and the commits can be merged if you choose to merge them! Thanks! :)

eeshangarg commented:
@timabbott: Updated! I couldn't change the order of the commits for some reason, even though locally I changed it and the last commit is actually the first on my local fork.

timabbott commented:
That's a longstanding GitHub display bug; the commits are in the right order in Git, but GitHub displays them in the incorrect order.  You can read all about it here: https://github.com/isaacs/github/issues/386

timabbott commented:
Looks great, merged, thanks @eeshangarg!  I'm not altogether happy with the way we have the Bitbucket webhook constructed, but we can revisit that in the future.

timabbott commented:
I did a little bit of reading, looking at our webhooks that have almost every line of code with test coverage (ideally we'd get them all to 100%).  It feels like 2/3 are basically versions of the "unsupported event type" case.

I think there's two valid strategies for these:
(1) Add a test case where we send a payload with an invalid event type, and do an `assertRaises` for those
(2) Decide we don't care enough to do that work, and just mark these lines as excluded from coverage.  The right way to do this is, rather than adding `#nocoverage` comments everywhere, to add a common exception type with a name like `UnexpectedWebhookEventType` that these webhooks can all raise, and put `raise UnexpectedWebhookEventType` in `tools/coveragerc`, just like `raise AssertionError` is, as a type of issue to ignore.

I think (2) is a pretty reasonable option, and would allow us to get 2/3 of our webhooks that don't have 100% test coverage to be cleared as "fully covered" without doing any real work, so I favor that approach.  What do you think?

eeshangarg commented:
@timabbott: I think the second option sounds much better than the first. Constructing "invalid" payloads for dozens of webhooks seems like too much work just for the sake of test coverage. Not to mention, every advanced webhook has dozens of events and there are a lot of different reasons we might choose not to support an event, so that makes me feel that this really doesn't need to be that big of a project. It makes much more sense to have a robust Exception-based framework for unsupported event types and build on top of that as necessary. :)

I definitely favor (2). I can get started on that fairly soon actually! Thanks, and sorry for the late response, didn't see this till now! :)
